sql/core/src/test/scala/org/apache/spark/sql/connector/WriteDistributionAndOrderingSuite.scala

```diff
@@ -0,0 +1,594 @@
+    val expectedPartitioning = partitioning match {
+      case p: physical.RangePartitioning =>
+        val resolvedOrdering = p.ordering.map(resolveAttrs(_, plan))
+        p.copy(ordering = resolvedOrdering.asInstanceOf[Seq[catalyst.expressions.SortOrder]])
+      case p: physical.HashPartitioning =>
+        val resolvedExprs = p.expressions.map(resolveAttrs(_, plan))
+        p.copy(expressions = resolvedExprs)
+      case other => other
+    }
+    // TODO: can be compatible, does not have to match 100%
+    assert(actualPartitioning == expectedPartitioning, "partitioning must match")
+
+    val actualOrdering = plan.outputOrdering
+    val expectedOrdering = ordering.map(resolveAttrs(_, plan))
+    // TODO: can be compatible, does not have to match 100%
+    assert(actualOrdering == expectedOrdering, "ordering must match")
+  }
+
+  private def resolveAttrs(
+      expr: catalyst.expressions.Expression,
+      plan: SparkPlan): catalyst.expressions.Expression = {
+
+    expr.transform {
+      case UnresolvedAttribute(parts) =>
+        val attrName = parts.mkString(",")
+        plan.output.find(a => a.name == attrName).get
+    }
+  }
+
+  private def attr(name: String): UnresolvedAttribute = {
+    UnresolvedAttribute(name)
+  }
+
+  private def catalog: ExtendedInMemoryTableCatalog = {
+    val catalog = spark.sessionState.catalogManager.catalog("testcat")
+    catalog.asTableCatalog.asInstanceOf[ExtendedInMemoryTableCatalog]
+  }
+
+  // executes a write operation and keeps the executed physical plan
+  private def execute(writeFunc: => Unit): SparkPlan = {
+    var executedPlan: SparkPlan = null
+
+    val listener = new QueryExecutionListener {
+      override def onSuccess(funcName: String, qe: QueryExecution, durationNs: Long): Unit = {
+        executedPlan = qe.executedPlan
+      }
+      override def onFailure(funcName: String, qe: QueryExecution, exception: Exception): Unit = {
+      }
+    }
+    spark.listenerManager.register(listener)
+
+    writeFunc
+
+    sparkContext.listenerBus.waitUntilEmpty()
+
+    executedPlan.asInstanceOf[WriteToDataSourceV2Exec].query
+  }
+}
+
+class ExtendedInMemoryTableCatalog extends InMemoryTableCatalog {
+
+  import org.apache.spark.sql.connector.catalog.CatalogV2Implicits._
+
+  def createTable(
+      ident: Identifier,
+      schema: StructType,
+      partitions: Array[Transform],
+      properties: util.Map[String, String],
+      distribution: Distribution,
+      ordering: Array[SortOrder]): Table = {
+
+    if (tables.containsKey(ident)) {
+      throw new TableAlreadyExistsException(ident)
+    }
+
+    InMemoryTableCatalog.maybeSimulateFailedTableCreation(properties)
+
+    val table = new ExtendedInMemoryTable(
+      s"$name.${ident.quoted}", schema, partitions, properties, distribution, ordering)
+    tables.put(ident, table)
+    namespaces.putIfAbsent(ident.namespace.toList, Map())
+    table
+  }
+}
+
+class ExtendedInMemoryTable(
```

Comment:
Why not add the distribution and order to the existing `InMemoryTable` and set default args that disable sorting and distribution?